distributed/protocol/__init__.py

```diff
@@ -117,3 +117,47 @@ def _register_cudf():
 def _register_cuml():
     with ignoring(ImportError):
         from cuml.comm import serialize
+
+
+import collections.abc
+
+import dask
+
+
+class TypeCompressor(collections.abc.MutableMapping):
+    def __init__(self, data=None):
+        self.storage = {}
+        if data is not None and isinstance(data, collections.abc.MutableMapping):
+            for key, value in data:
+                header, frames = serialize(value)
+                self.storage[key] = (header, maybe_compress(frames))
+
+    def __setitem__(self, key, value):
+        header, frames = serialize(value)
+        self.storage[key] = (header, maybe_compress(frames))
+
+    def __getitem__(self, key):
+        header, (compression, compressed) = self.storage[key]
+
+        frames = decompress({"compression": {compression}}, compressed)
+        return deserialize(header, frames)
```

Comment:
In principle this looks good @prasunanand .  Thank you for sticking with this.

The next thing to do is probably to try it out in practice a bit and see how it performs.  There are optimizations I can think of, like it might make sense to only compress values that we know are small using the `sizeof` function, but that would just be a guess.  If you have time to try out a few different kinds of computations, maybe taken from examples.dask.org that might be a good way to get more information here.

Reply:
@prasunanand , have you had any time to try the suggestions here?

Reply:
I thought this is what `maybe_compress` already does. Meaning it checks to see whether something is worth compressing. If so, it compresses it.